New CSS property: rule (#44947)
mdn/content: files/en-us/web/css/reference/properties/rule/index.md

---
title: "`rule` CSS property"
short-title: rule
slug: Web/CSS/Reference/Properties/rule
page-type: css-property
status:
  - experimental
browser-compat: css.properties.rule
sidebar: cssref
---

{{SeeCompatTable}}

> The **`rule`** [shorthand](/en-US/docs/Web/CSS/Guides/Cascade/Shorthand_properties) [CSS](/en-US/docs/Web/CSS) property sets the width, style, and color of the line drawn between rows and columns in multi-row grid, flex, setting both the column and row rules to the same values.

Reviewer: ```suggestion
The **`rule`** [shorthand](/en-US/docs/Web/CSS/Guides/Cascade/Shorthand_properties) [CSS](/en-US/docs/Web/CSS) property sets the width, style, and color of the line drawn between rows and columns in multi-row grid, flex, and multi-col layouts, setting both the column and row rules to the same values.
```

multi-col relevant too, right?

{{InteractiveExample("CSS Demo: rule")}}

```css interactive-example-choice
rule: solid;
```

```css interactive-example-choice
rule: dotted medium blue;
```

```css interactive-example-choice
rule:
  dotted medium blue,

>   repeat(3, dotted purple 1px, double orange 5px);

Reviewer: ```suggestion
  repeat(3, dotted red 2px, double orange 5px);
```

The 1px purple is really hard to see on the black background.

```

```css interactive-example-choice
rule:
  dashed medium magenta,

>   repeat(auto, dotted blue 1px, dotted blue 5px),

Reviewer: ```suggestion
  repeat(auto, dotted blue 2px, dotted blue 5px),
```

Again, just make the thinner blue lines slightly easier to see?

  dashed medium magenta;
```

```css interactive-example-choice
rule:
  dashed medium magenta,

>   repeat(auto, dotted blue 1px),

Reviewer: ```suggestion
  repeat(auto, dotted blue 2px),
```

Same again?

  outset goldenrod 5px;
```

```html interactive-example
<section id="default-example">
  <div id="example-element">
    <i>A</i>
    <i>B</i>
    <i>C</i>
    <i>D</i>
    <i>E</i>
    <i>F</i>
    <i>G</i>
    <i>H</i>
    <i>I</i>
    <i>J</i>
    <i>K</i>
    <i>L</i>
    <i>M</i>
    <i>N</i>
    <i>O</i>
    <i>P</i>
    <i>Q</i>
    <i>R</i>
    <i>S</i>
    <i>T</i>
    <i>U</i>
    <i>V</i>
    <i>W</i>
    <i>X</i>
    <i>Y</i>
    <i>Z</i>
  </div>
</section>
```

```css interactive-example
#example-element {
  display: grid;
  grid-template-columns: repeat(7, 1fr);
  rule: solid thick;
}
#example-element i {
  padding: 5px;
}
```

## Constituent properties

This property is a shorthand for the following CSS properties:

- {{cssxref("rule-color")}}
- {{cssxref("rule-style")}}
- {{cssxref("rule-width")}}

## Syntax

```css
/* One value */
rule: dotted;
rule: solid 8px;
rule: solid blue;
rule: thick inset blue;

/* Multiple values */
rule: groove, dashed, solid;
rule:
  dotted medium blue,
  dashed magenta 1px,
  outset green 5px;
rule:
  solid #0ff,
  repeat(3, dashed magenta 1px, outset green 5px);

> rule:

Reviewer: Maybe provide one example with a value featuring a `<gap-rule>`, followed by a `repeat()` function, followed by another `<gap-rule>`? Just in case people think the `repeat()` function has to come last?

  inset 3px yellow,

>   repeat(auto, dashed magenta 1px, groove green 5px);

Author: ```suggestion
  repeat(auto, dashed magenta 1px, groove green 5px),
  inset 3px yellow;
```

/* Global values */
rule: inherit;
rule: initial;
rule: revert;
rule: revert-layer;
rule: unset;
```

### Values

The `rule` property accepts a comma-separated list of values, including:

- `<gap-rule>`
  - : Specified as one, two, or three of the values listed below, in any order.
    - `<'line-width'>`

>       - : A {{cssxref("&lt;line-width&gt;")}}: a positive {{cssxref("&lt;length&gt;")}} or one of the three keywords, `thin`, `medium`, or `thick`. The default value is `medium. See {{cssxref("rule-width")}}.

Reviewer: ```suggestion
      - : A {{cssxref("&lt;line-width&gt;")}}: a positive {{cssxref("&lt;length&gt;")}} or one of the three keywords, `thin`, `medium`, or `thick`. The default value is `medium`. See {{cssxref("rule-width")}}.
```

```css
rule: dashed red 3px;
```

When more than one `<gap-rule>` is declared, they will be applied to rules in the order specified. If there are more gutters between rows and columns than `<gap-rule>` values, the list of values is repeated until every row and column rule has a gap line. If we declare the following, for example, every odd rule will be `dashed red 3px`, and every even rule will be `dotted blue 5px`, in both directions.

```css
rule:
  dashed red 3px,
  dotted blue 5px;
```

### Repeated line styles

The `repeat()` function, with an integer of `1` or greater as the first argument, can be used to repeat a valid list of CSS [`<gap-rule>`](#gap-rule) values passed as subsequent arguments the specified number of times. This allows the same `<gap-rule>` to be repeated a set number of times without repeating the same CSS multiple times. The following declarations are equivalent:

```css
rule:
  solid red 5px,
  outset blue 10px,
  inset green 1px,
  outset blue 10px,
  inset green 1px,
  outset blue 10px,
  inset green 1px;
rule:
  solid red 5px,
  repeat(3, outset blue 10px, inset green 1px);
```

> This creates a list of seven rules. If the number of styles in the `rule` value's style list exceeds the number of gaps between rows and columns, the excess style values are ignored. If the container on which this is applied has three rows and columns, the rule in the first gutter will be `solid red 5px` and the second `outset blue 10px`.

Reviewer: ```suggestion
This creates a list of seven rules. If the number of styles in the `rule` value's style list exceeds the number of gaps between rows and columns, the excess style values are ignored. If the container on which this is applied has three rows and columns, the rule in the first gutter will be `solid red 5px` and the second `outset blue 10px`, in both directions.
```

```css live-sample___repeat
ul {
  rule:
    red dashed,
    1px dotted,
    5px blue,
    10px magenta solid;
}
```

{{EmbedLiveSample("Repeat", "", "600")}}

The red line is `3px` wide, the dotted line is the same color as the text, and there is no `5px`-wide blue line, as the style of the third `<gap-rule>` defaults to `none`, so no line is painted. As there are fewer rule styles than gutters, the list of rules is repeated until all rule lines are styled.

### Using the `repeat()` function

This example demonstrates using the `repeat()` function within the `rule` property value. We use the same HTML and CSS as in the previous examples, overriding the `rule` value with a comma-separated list of three components: two `<gap-rule>` values and a `<gap-repeat-rule>` that sets a list of two `<gap-rule>` values to repeat 3 times.

```css live-sample___func live-sample___auto
ul {
  rule:
    3px red dashed,
    repeat(3, dotted green 1px, dashed blue 1px),
    3px red dashed;
}
```

{{EmbedLiveSample("func", "", "600")}}

> The grid has ten columns and eight rows, so nine column gutters and seven row gutters. The `repeat()` function repeats two style values three times, creating a list of eight style values. As there are fewer row gutters than value, the last value is not used in the row direction. As there are more column gutters that rule style values, the list is repeated in the column direction.

Reviewer: ```suggestion
The grid has ten columns and eight rows, so nine column gutters and seven row gutters. The `repeat()` function repeats two style values three times, creating a list of eight style values. As there are fewer row gutters than values, the last value is not used in the row direction. As there are more column gutters than values, the list repeats in the column direction.
```

### Using `auto` within `repeat()`

This example demonstrates using the `auto` argument instead of an integer in the `repeat()` function.

> Using `repeat(auto, <gap-rule>)`, we set all row and column rules to `1px dotted` and default to the current color, except the first and last rules, which we set to `3px solid red`.

Reviewer: ```suggestion
Using `repeat(auto, <gap-rule>)`, we set all row and column rules to `1px dotted` (with the color defaulting to the current color), except the first and last rules, which we set to `3px solid red`.
```
